bugsnag-android-core/src/test/java/com/bugsnag/android/DeviceDataSerializationTest.kt

```diff
@@ -17,19 +21,49 @@ internal class DeviceDataSerializationTest {
         @JvmStatic
         @Parameters
         fun testCases(): Collection<Pair<Map<String, Any>, String>> {
-            val context = Mockito.mock(Context::class.java)
-            val res = Mockito.mock(Resources::class.java)
-            val prefs = Mockito.mock(SharedPreferences::class.java)
-            val editor = Mockito.mock(SharedPreferences.Editor::class.java)
-            Mockito.`when`(prefs.edit()).thenReturn(editor)
-            Mockito.`when`(editor.putString(Mockito.anyString(), Mockito.anyString())).thenReturn(editor)
+            val context = mock(Context::class.java)
+            val res = mock(Resources::class.java)
+            val conf = mock(Configuration::class.java)
+            val connectivity = mock(Connectivity::class.java)
+
+            val prefs = mock(SharedPreferences::class.java)
+            val editor = mock(SharedPreferences.Editor::class.java)
+            `when`(prefs.edit()).thenReturn(editor)
+            `when`(editor.putString(anyString(), anyString()))
+                .thenReturn(editor)
 
             val buildInfo = DeviceBuildInfo(
                 "samsung", "s7", "7.1", 24, "bulldog",
                 "foo-google", "prod,build", "google", arrayOf("armeabi-v7a")
             )
-            val deviceData = DeviceData(null, context, res, "123", buildInfo)
-            return generateSerializationTestCases("device_data", deviceData.deviceDataSummary)
+
+            // regular fields
+            `when`(res.configuration).thenReturn(conf)
+            conf.orientation = 2
+
+            // metadata fields
+            val metrics = DisplayMetrics()
+            metrics.density = 200f
+            metrics.densityDpi = 120
+            `when`(res.displayMetrics).thenReturn(metrics)
+            `when`(connectivity.retrieveNetworkAccessState()).thenReturn("unknown")
+
+            // construct devicedata object
+            val deviceData = DeviceData(connectivity, context, res, "123", buildInfo, File(""))
+
+            // serializes the 3 different maps that DeviceData can generate:
+            // 1. summary (used in session payloads)
+            // 2. regular (used in event payloads)
+            // 3. metadata (used in event payloads)
+            val regularMap = deviceData.deviceData
+            arrayOf("freeDisk", "freeMemory", "totalMemory", "locale", "time").forEach { regularMap.remove(it) }
```

Comment:
Why are these removed from the test? Can't we mock the data providers?

Reply:
freeMemory/totalMemory cannot be mocked in their current form as they rely on the `Runtime` class directly. We could add a wrapper class that decorates all calls to `Runtime` which would be mocked if this is desirable.

`freeDisk` requires a mock `File`, which could be added.

`locale` requires a mock/fake `Locale`, which would require adding another constructor parameter.

`time` requires passing in a `Date` as a constructor parameter.

If you're happy with making these alterations to the class then I can update the test accordingly.

Reply:
OK - let's address the existing testability constraints separately at the end of the refactor.

However, in this case I don't see how the changes to pass in dataDirectory have gained us anything as this isn't being mocked in the test anyway. Unless I'm missing something, I believe this would be the same regardless of whether the directory was set internally or externally.

Suggest we remove this for now, or if we are keeping the parameter to the constructor then I don't see a need to have the parameter to `calculateFreeDisk` as it can use the member variable (otherwise it's effectively a rather pointless static function).

Reply:
`Environment.getDataDirectory()` returns null when run in a JVM test, which causes a NPE that fails the whole test case. I'd therefore suggest one additional change to this PR of mocking the parameter and then address whether we need to use PowerMock/other mocking frameworks at the end of the refactor, if you're happy with that approach.

Reply:
Yes, as long as `calculateFreeDisk` is also corrected to use member variable, I'm happy with that.